### PR TITLE
Unrecognized even tags are unbound and assigned negative inscription numbers

### PR DESCRIPTION
**Pull request by raphjaph**

### CODE REVIEW

**Comment by ddiekroeger** [comment]
what does "unbound" mean?

**Comment by dannydeezy** [comment]
wait no no - we so unrecognized even tag means the inscription is assigned to a fake UTXO ??

**Comment by raphjaph** [comment]
We introduced the concept of unbound in commit https://github.com/casey/ord/commit/950740aebc28891424e590fd6eb2a50536cea6f1 because of Super_Testnet's zero-value transaction. Even tags have always been reserved for core protocol things and right now we ignore inscriptions that use them but at the same time we want to recognize as many inscriptions as possible so this is what we came up with.

**Comment by dannydeezy** [comment]
> 

thanks, so unbound means "not assigned to a sat" and therefore not transferrable right?

**Comment by raphjaph** [comment]
Yes, that is correct. 

We are doing this because an even tag might be used in the future to point to a specific sat in the reveal tx outputs so as to assign an inscription to a different than the default sat. If people are overloading the even tags with their own meaning this might bite us in the ass in the future.

**Comment by dannydeezy** [comment]
Ok that makes sense. However, currently a lot of people are deliberately trying to inscribe "cursed" with an OP_66 tag, and all these people will feel rugged if they don't own their inscription.

I propose that since a large portion of the community has adopted the OP_66 cursed tag, that we recognize OP_66 (and just take no action, but this ensures the inscription gets assigned normally). I can submit a PR.

**Comment by dannydeezy** [comment]
@raphjaph i made a PR to this branch that would recognize `OP_66` as "manually cursed" and prevent it from being unbound

https://github.com/raphjaph/ord/pull/10/files

**Comment by casey** [comment]
> Ok that makes sense. However, currently a lot of people are deliberately trying to inscribe "cursed" with an OP_66 tag, and all these people will feel rugged if they don't own their inscription.

I actually think this is a pretty strong reason *not* to add support for OP_66. Issues, draft PRs, and discussing future plans for the protocol are not guarantees that specific changes will be made. It is often the case that during the course of development, plans change, and something that once seemed like a good idea turns out to have unforeseen consequences, causing plans to change.

So, users are of course free to make nonstandard inscriptions, such as OP_66 cursed inscriptions, in the hopes that `ord` will treat them in a particular way in the future. However, `ord` may not, and users should be prepared for such an eventuality. Learning that lesson now seems better than later.

**Comment by dannydeezy** [comment]
> I actually think this is a pretty strong reason _not_ to add support for OP_66. Issues, draft PRs, and discussing future plans for the protocol are not guarantees that specific changes will be made. It is often the case that during the course of development, plans change, and something that once seemed like a good idea turns out to have unforeseen consequences, causing plans to change.
> 
> So, users are of course free to make nonstandard inscriptions, such as OP_66 cursed inscriptions, in the hopes that `ord` will treat them in a particular way in the future. However, `ord` may not, and users should be prepared for such an eventuality. Learning that lesson now seems better than later.

i think the lesson is learned already (I personally feel very terrible for causing this trouble) but denying `OP_66` would make a lot of regular people unhappy

**Comment by dannydeezy** [comment]
btw, are there links to discussion around the decision to make unrecognized even tags unbound?

**Comment by DrJingLee** [comment]
Seems PPL are happy to make mistakes..However PPL love the idea of “cursed”and “blessed” inscription...the cursed is as cool aa “darkenergy ”

**Comment by bitcoinunleashed** [comment]
> > Ok that makes sense. However, currently a lot of people are deliberately trying to inscribe "cursed" with an OP_66 tag, and all these people will feel rugged if they don't own their inscription.
> 
> I actually think this is a pretty strong reason _not_ to add support for OP_66. Issues, draft PRs, and discussing future plans for the protocol are not guarantees that specific changes will be made. It is often the case that during the course of development, plans change, and something that once seemed like a good idea turns out to have unforeseen consequences, causing plans to change.
> 
> So, users are of course free to make nonstandard inscriptions, such as OP_66 cursed inscriptions, in the hopes that `ord` will treat them in a particular way in the future. However, `ord` may not, and users should be prepared for such an eventuality. Learning that lesson now seems better th

But you haven't presented any evidence that you have discovered technical consequences to implementing this. The only reason you are giving above is that people need to learn. So plans are not changing for any reason other than Casey wants people to learn a lesson by losing money, correct?

**Comment by jokie88** [comment]
I strongly support prioritizing the long-term success of the Ordinals protocol over speculators who have rushed into an evolving situation. We degens are well aware of the risks involved when being early apers. While I don't want to lose money, the potential loss will be even greater if Ordinals fails in the long run.

The success of the Ordinals protocol is most important, and I believe simplicity is crucial for its longevity. It's easier to add complexity, but much harder to remove it. We are already experiencing significant complexity in trying to preserve the inscription numbers, with the indexer needing to adapt at different block heights.

Every additional layer of complexity will further complicate matters. I now lean towards the belief that inscription numbers should be flexible to a certain extent, and we shouldn't expect them to be permanent, especially in these early stages of the Ordinals journey.

100 year vision!

**Comment by dannydeezy** [comment]
a 100-year outlook harmonious solution should aim to keep the community of users, developers, and degens aligned

**Comment by spencerrichardhenry** [comment]
A 100-year outlook harmonous solution should re-index all negative inscriptions in their proper place, not hack it into new negative numbers for fear of upsetting a few people.

**Comment by drahlraigon** [comment]
The Universe has place for imperfections and your 100 year outlook harmonous solution doesn’t? @spencerrichardhenry 
Also it is not a question of a few people anymore, as we speak around 70k negative insciptions have been inscribed, millions of dollars in fees (both significant numbers, definitely not few)

**Comment by jokie88** [comment]
> The Universe has place for imperfections and your 100 year outlook harmonous solution doesn’t? @spencerrichardhenry Also it is not a question of a few people anymore, as we speak around 70k negative insciptions have been inscribed, millions of dollars in fees (both significant numbers, definitely not few)

I've heard there are multiple groups w/ cursed inscribers that do not have this op66 tag issue. my guess is the impact is much smaller. Additionally -- 70k over a few days is nothing compared to the numbers of ordinals this codebase will have to support decades into the future. we already have 7m in a few months.

**Comment by drahlraigon** [comment]
> > The Universe has place for imperfections and your 100 year outlook harmonous solution doesn’t? @spencerrichardhenry Also it is not a question of a few people anymore, as we speak around 70k negative insciptions have been inscribed, millions of dollars in fees (both significant numbers, definitely not few)
> 
> I've heard there are multiple groups w/ cursed inscribers that do not have this op66 tag issue. my guess is the impact is much smaller. Additionally -- 70k over a few days is nothing compared to the numbers of ordinals this codebase will have to support decades into the future. we already have 7m in a few months.

Yes but what exactly is the problem in the codebase supporting these inscriptions decades in the future haha. The way i see it, codes have bugs- and by keeping the limited no of negative inscriptions that have already been inscribed we are just celebrating history and the beauty of imperfections that exist in every realm of life, ordinals being no exception.

**Comment by scooty69** [comment]
Negative inscriptions should remain imho. At first, it might have been OK to eliminate them in various methods - at this point, there’s clear demand for it and I think it present an interesting wrinkle to our space. Eliminating them now would harm many regular people and harm the reputation of the protocol by association.

most interestingly to me, negative numbers allow for more interesting artistic opportunities, such as we’ve seen with people inscribing negative colors on negative numbers, or collections that may have brother/sister inscriptions on both positive and negative. We’ve only scratched the surface of how this could be used in the future and squashing it now will squash amazing future discoveries and experiments not yet made. 

We should be pushing the limits of what’s possible, not allowing only sanctioned experiments.

**Comment by tyjvazum** [comment]
I'm willing to support unrecognized even tags being transferable in [arb](https://github.com/tyjvazum/arb) by removing the current restriction without making such inscriptions unbound. This can be implemented relatively quickly. Future upgrades using even tags will still be possible by using an activation height system. Is anyone interested in this solution?

**Comment by letsberational** [comment]
I think before making any decisions, we should actually be looking at the hard work creators have put in with cursed ordinals.
Its easy simply brush them off before looking at the artwork that have been inspired by the cursed world.

I made my first original ordinal collection based around the concept that my ordinals were cursed and belonged in the underworld.

For them to be kicked out of the negative ordinal ecosystem, and make them re-enter the positive world, would destroy these cute cursed characters.

There are other artists who have launched on cursed ordinals, which should also highlight the work they've done.

Inscription-4065
Inscription-4029
Inscription-5874
Inscription-6896

**Comment by dannydeezy** [comment]
@raphjaph @casey what if we did something like this:
- if there are any unrecognized even tags it becomes cursed
- if there are any unrecognized tags divisible by 4 it is made unbound

gives us flexibility around the decision of bound vs. unbound for future stuff

**Comment by CyberApeKing** [comment]
There have been many instances of "bugs", where people could duplicate to get extra items or someone could do something to get xyz. The bug gets patched and people keep whatever they had as this was a bug from the companies side - the people aren't at fault. 

What's special in this scenario? People have made amazing art in the process, with cursed themes, whole communities formed and clearly a lot of interest and time invested. Great history for Ordinals, where people turned a bug into something amazing.

I suggest patching negatives ASAP and keeping whatever items people managed to obtain as has been historically. Fun, unique Ordinal event that has a good story and had a canvas for creativity.

**Comment by WalletDatGH** [comment]
It is a new method of inscriptions that has come out of nowhere, it does not harm and in my opinion it follows the bitcoin philosophy, it does not have a controlling organization, in this case it is not even created on purpose. There is a community behind that wants to continue with the inscriptions that it has mined, also in collecting errors do not have to be a problem, on the contrary, as long as it does not harm it, it is still something valuable.

**Comment by bootosh** [comment]
There's currently a cursed inscription community forming of <-10k and people really feel involved in a movement, it's like a secondary home for those who missed the original <10k wave. Especially with how it was talked about in major twitter spaces a lot of people didn't even know that they can get burnt from this. Just gave the appearance of "cool new way to inscribe"

Seconding on some cool art things that people made, such as the inverse cult/more "cursed" wave of art and other neat things. I am biased but I also think they should remain if possible

**Comment by buxor** [comment]
Subzero
Let's find a way for negative inscriptions to exist as cursed ordinals, otherwise sub-zero will come get us all.

**Comment by nick07002** [comment]
mbpr

Mesh Beatles support cursed ordinals. Fun brings people together on BTC, isn't that's what we want?
More activities, more stable BTC network.

**Comment by IHAXONI** [comment]
It's like traveling back in time, this is the only chance, make it happen.

**Comment by honeyrohada** [comment]
This should be given a green signal because people are in favor to do so. more like _**expanding the BTC creations and breaking out from limits.**_

**Comment by sanj770** [comment]
> @raphjaph @casey what if we did something like this:
> 
> * if there are any unrecognized even tags it becomes cursed
> * if there are any unrecognized tags divisible by 4 it is made unbound
> 
> gives us flexibility around the decision of bound vs. unbound for future stuff



I like the proposal from @dannydeezy 

If we're able to still maintain flexibility going forwards and also allow the current cursed inscriptions to be recognised I definitely think that would be what a lot of people would be in favour of.

It's also important to give @casey @raphjaph and the team time to think it over properly and not rush due to the crazy speculation at the moment!

Finding the best option for the longterm success of the Ordinals protocol as a whole is what’s most important! Hopefully this includes cursed inscriptions as it’s reached a point where a large number of people have invested. 

I feel like the ‘cursed saga' could be a pretty historical moment to look back on in the future as to how it was handled as it’s one of the first times consensus might be tested.

**Comment by tyjvazum** [comment]
@raphjaph, with @casey quitting and handing this repo over to you, this is a pivotal moment. Let's work together to solve things in a way that respects the millions of dollars in transaction fees spent by users expecting to create NFTs, not [donations to a museum](https://twitter.com/veryordinally/status/1661765514853482496). All inscriptions should be treated as valid inscriptions and be transferable. Let's start working together as a community to respect people's property rights, as is the Bitcoin way.

**Comment by ehutzle** [comment]
Every single person I've talked to regarding cursed ordinals supports keeping them bound. In fact, cursed ordinals served as an onboarding ramp to new folks, and folks who haven't inscribed.

Let's keep them bound.

File: src/index/updater/inscription_updater.rs [hunk 1]
```diff
@@ -111,30 +127,24 @@ impl<'a, 'db, 'tx> InscriptionUpdater<'a, 'db, 'tx> {
       }
     }
 
-    if inscriptions.iter().all(|flotsam| flotsam.offset != 0)
-      && Inscription::from_transaction(tx).is_some()
-    {
+    let (inscription, cursed, unbound) = match Inscription::from_transaction(tx) {
```
Comment (raphjaph) [inline src/index/updater/inscription_updater.rs:130]: return true instead of inscription

File: src/index.rs [hunk 2]
```diff
@@ -2279,4 +2279,109 @@ mod tests {
       );
     }
   }
+
```
Comment (raphjaph) [inline src/index.rs:2282]: add test for zero value transactions

**Review by casey — APPROVED** [review]
Looks good to me!

**Comment by casey** [comment]
Although we should remove the inscription from the error message and just return a bool instead of `Option<Inscription>`

**Comment by raphjaph** [comment]
As the protocol currently stands inscriptions are not valid if they use an unrecognized even tag, so this change already makes a concession by recognizing them. For now they are unbound but we might reconsider this and bind them in the future if there are strong reasons.

**Comment by tyjvazum** [comment]
> As the protocol currently stands inscriptions are not valid if they use an unrecognized even tag, so this change already makes a concession by recognizing them. For now they are unbound but we might reconsider this and bind them in the future if there are strong reasons.

Well, this issue went poorly, as expected. For anyone interested in a solution to recognize the millions of dollars in transaction fees spent by users intending to create NFTs, not [donations to a museum](https://twitter.com/veryordinally/status/1661765514853482496), I'm working on it [here](https://github.com/tyjvazum/arb/issues/16). I'm very open to outside discussion and collaboration, so I'll respond to community input instead of ignoring it 🙂

**Comment by satoshi0770** [comment]
support,
"That way the inscriptions aren't lost and we don't encourage people to make weird inscriptions in an attempt to get special treatment." -gmart7t2

**Comment by Frty47222** [comment]
A ‘Bug’ should be a ‘Feature’ too, Nothing is born perfect

**Comment by dannydeezy** [comment]
why is "unbound" the default?

**Comment by letsberational** [comment]
> As the protocol currently stands inscriptions are not valid if they use an unrecognized even tag, so this change already makes a concession by recognizing them. For now they are unbound but we might reconsider this and bind them in the future if there are strong reasons.

There are already strong reasons....over 50,000 strong reason, because that's how many cursed ordinals have been inscribed.
The demand is there, and the community is there. Why are the devs so against cursed ordinals?

**Comment by concernedordinal** [comment]
Hey @casey and @raphjaph I found out about this debate from the bitcoin magazine article. No cursed ordinals in my wallet yet but I really like the idea and works. I think they should be fully supported so they can be traded like true ordinals then maybe I can buy some. It would feel like a real rug pull if you go ahead with the choice to not back your community. Look at all the friendly degens myself included that want you to let them have their cursed ordinals. Even the news is covering it now. Please do the right thing and show that you work for the community you helped to build. That is the way forward.

https://bitcoinmagazine.com/technical/the-debate-around-cursed-ordinal-inscriptions

**Comment by BatBushRacks** [comment]
Still radio silence from the devs... crazy... did some digging looks like development quietly moved to https://github.com/raphjaph/ord/commits/cursed-inscriptions-the-beginning... not a dev so no idea what it mean... any improvement?

**Comment by veryordinally** [comment]
> Still radio silence from the devs... crazy... did some digging looks like development quietly moved to https://github.com/raphjaph/ord/commits/cursed-inscriptions-the-beginning... not a dev so no idea what it mean... any improvement?

We're tackling the cursed inscriptions step by step and have released initial support in 0.6.0. We will get to this issue soon. Your continued support is appreciated.

**Comment by BatBushRacks** [comment]
> > Still radio silence from the devs... crazy... did some digging looks like development quietly moved to https://github.com/raphjaph/ord/commits/cursed-inscriptions-the-beginning... not a dev so no idea what it mean... any improvement?
> 
> We're tackling the cursed inscriptions step by step and have released initial support in 0.6.0. We will get to this issue soon. Your continued support is appreciated.

@veryordinally super happy to continue supporting you and the other devs... glad you appreciate it... nothing much to get to though... are you aligned with the community on making our NFTs function as NFTs... or as @letsberational said... will you keep fighting the community?

**Comment by concernedordinal** [comment]
Hey @casey @raphjaph and @veryordinally another article was recently published about this nightmare. It says you plan on supporting these cursed inscriptions as true ordinals the same as everyone in the community wants! Why are you all purposely not saying that here then? Like @BatBushRacks said it should be easy and quick to state your intentions. Silence can be deafening. Just my view on things.

https://cointelegraph.com/news/bitcoin-ordinals-upgrade-rectify-cursed-inscriptions-issue

**Comment by creamplanet** [comment]
Greetings all, just read through everything. I think these should be handled the same as the other inscriptions, but at the same time everyone should remember open source is unpaid work. No one "works for the community" on this.

**Comment by BatBushRacks** [comment]
> Greetings all, just read through everything. I think these should be handled the same as the other inscriptions, but at the same time everyone should remember open source is unpaid work. No one "works for the community" on this.

Well they ask for and have received donations from the community... https://twitter.com/rodarmor/status/1662617512700420096... so in that way it is working for the community... the job is to make our NFTs function... an important job indeed...

**Comment by emmaeverleigh** [comment]
I read about this in the BM article. I feel they should be recognised but are inscription numberss really a good idea? I feel like a lot of drama has been brought up by the numbers while watching the Ordinals space the past few months.  A lot of people say they should be dropped. Anyway thanks to the team for building all of this.

**Comment by raphjaph** [comment]
#2359
